### PR TITLE
feat: Add support to install che using stable-all-namespaces olm channel

### PR DESCRIPTION
**Pull request by flacatus**
Signed-off-by: Flavius Lacatusu <flacatus@redhat.com>

### What does this PR do?
Add support for stable-all-namespaces channel.

https://github.com/eclipse-che/che-operator/pull/890

### What issues does this PR fix or reference?

Issue: https://github.com/eclipse/che/issues/19891

### PR Checklist

[As the author of this Pull Request I made sure that:](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#pull-request-template-and-its-checklist)

- [x] [Code produced is complete](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-produced-is-complete)
- [x] [Code builds without errors](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-builds-without-errors)
- [ ] [Tests are covering the bugfix](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#tests-are-covering-the-bugfix)
- [ ] [The repository devfile is up to date and works](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-repository-devfile-is-up-to-date-and-works)
- [ ] [Sections `What issues does this PR fix or reference` and `How to test this PR` completed](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#sections-what-issues-does-this-pr-fix-or-reference-and-how-to-test-this-pr-completed)
- [ ] [Relevant user documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [Relevant contributing documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [CI/CD changes implemented, documented and communicated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#cicd-changes-implemented-documented-and-communicated)

### Reviewers

Reviewers, please comment how you tested the PR when approving it.

### CODE REVIEW

File: src/api/context.ts [hunk 1]
```diff
@@ -45,11 +45,15 @@ export namespace ChectlContext {
   export async function init(flags: any, command: Command): Promise<void> {
     ctx.isChectl = getProjectName() === CHECTL_PROJECT_NAME
     ctx.isDevVersion = getProjectVersion().includes('next') || getProjectVersion() === CHECTL_DEVELOPMENT_VERSION
-
+    ctx.operatorNamespace = DEFAULT_CHE_NAMESPACE
```
Comment (tolusha) [inline src/api/context.ts:48]:
```suggestion
    ctx.operatorNamespace = flags.chenamespace || DEFAULT_CHE_NAMESPACE
```

File: src/tasks/installers/olm.ts [hunk 2]
```diff
@@ -313,43 +318,53 @@ export class OLMTasks {
           task.title = `${task.title}: ${ctx.isPreInstalledOLM}...OK`
         },
       },
+      {
+        title: `Check if operator is installed in ${DEFAULT_OPENSHIFT_OPERATORS_NS_NAME} namespace`,
```
Comment (tolusha) [inline src/tasks/installers/olm.ts:322]: Why do we need this?

Comment (flacatus) [reply src/tasks/installers/olm.ts:322]: I didn't found a better way to get the ctx.operatorNamespace. I will dig

File: src/commands/server/delete.ts [hunk 3]
```diff
@@ -73,13 +73,21 @@ export default class Delete extends Command {
...
+    const checlusters = await kube.getAllCheClusters()
+
     tasks.add(apiTasks.testApiTasks(flags))
     tasks.add(operatorTasks.deleteTasks(flags))
-    tasks.add(olmTasks.deleteTasks(flags))
     tasks.add(cheTasks.deleteTasks(flags))
     tasks.add(helmTasks.deleteTasks(flags))
     tasks.add(cheTasks.waitPodsDeletedTasks())
 
+    // Remove 'openshift-operators' subscription in case if exists at least 1 checluster
```
Comment (tolusha) [inline src/commands/server/delete.ts:84]: I don't like it.
I don't understand how subscription checking relates to all olmTasks

File: src/tasks/installers/olm.ts [hunk 4]
```diff
@@ -142,20 +144,23 @@ export class OLMTasks {
...
+              const catalogSourceNamespace = flags['catalog-source-namespace'] || ctx.operatorNamespace
+              subscription = this.constructSubscription(SUBSCRIPTION_NAME, flags['package-manifest-name'], ctx.operatorNamespace, catalogSourceNamespace, flags['olm-channel'], ctx.sourceName, ctx.approvalStarategy, ctx.startingCSV)
             } else if (VersionHelper.isDeployingStableVersion(flags) || flags['olm-channel'] === OLM_STABLE_CHANNEL_NAME) {
               // stable Che CatalogSource
-              subscription = this.constructSubscription(SUBSCRIPTION_NAME, DEFAULT_CHE_OLM_PACKAGE_NAME, flags.chenamespace, ctx.defaultCatalogSourceNamespace, OLM_STABLE_CHANNEL_NAME, ctx.catalogSourceNameStable, ctx.approvalStarategy, ctx.startingCSV)
+              subscription = this.constructSubscription(SUBSCRIPTION_NAME, DEFAULT_CHE_OLM_PACKAGE_NAME, ctx.operatorNamespace, ctx.defaultCatalogSourceNamespace, OLM_STABLE_CHANNEL_NAME, ctx.catalogSourceNameStable, ctx.approvalStarategy, ctx.startingCSV)
+            } else if (flags['olm-channel'] === STABLE_ALL_NAMESPACES_CHANNEL_NAME) {
```
Comment (tolusha) [inline src/tasks/installers/olm.ts:158]: Can we remove this extra `if` ?

File: src/tasks/installers/olm.ts [hunk 5]
```diff
@@ -202,14 +207,14 @@ export class OLMTasks {
       {
         title: 'Prepare Eclipse Che cluster CR',
         task: async (ctx: any, task: any) => {
-          const cheCluster = await kube.getCheCluster(flags.chenamespace)
+          const cheCluster = await kube.getCheCluster(ctx.operatorNamespace)
```
Comment (tolusha) [inline src/tasks/installers/olm.ts:210]: hmmm? should checluster goes into eclipse-che namespace?

File: src/tasks/installers/olm.ts [hunk 6]
```diff
@@ -202,14 +207,14 @@ export class OLMTasks {
...
           if (cheCluster) {
             task.title = `${task.title}...It already exists..`
             return
           }
 
           if (!ctx.customCR) {
-            ctx.defaultCR = await this.getCRFromCSV(kube, flags.chenamespace)
+            ctx.defaultCR = await this.getCRFromCSV(kube, ctx.operatorNamespace)
```
Comment (tolusha) [inline src/tasks/installers/olm.ts:217]: The same as above

**Comment by flacatus** [comment]
Not ready for review

File: src/tasks/installers/olm.ts [hunk 7]
```diff
@@ -174,22 +179,22 @@ export class OLMTasks {
...
+          await kube.waitUntilOperatorIsInstalled(ctx.installPlanName, ctx.operatorNamespace)
           task.title = `${task.title}...done.`
         },
       },
       {
         title: 'Set custom operator image',
         enabled: () => flags['che-operator-image'],
-        task: async (_ctx: any, task: any) => {
-          const csvList = await kube.getClusterServiceVersions(flags.chenamespace)
+        task: async (ctx: any, task: any) => {
```
Comment (AndrienkoAleksandr) [inline src/tasks/installers/olm.ts:196]: Sorry, here is you renamed _ctx to ctx, but didn't used this variable. Is it ok?

File: src/tasks/installers/olm.ts [hunk 8]
```diff
@@ -313,43 +318,61 @@ export class OLMTasks {
...
+        },
+      },
       {
         title: `Delete(OLM) operator subscription ${SUBSCRIPTION_NAME}`,
         enabled: ctx => ctx.isPreInstalledOLM,
-        task: async (_ctx: any, task: any) => {
-          await kube.deleteOperatorSubscription(SUBSCRIPTION_NAME, flags.chenamespace)
+        task: async (ctx: any, task: any) => {
+          const checlusters = await kube.getAllCheClusters()
+          if (kube.operatorSubscriptionExists(SUBSCRIPTION_NAME, DEFAULT_OPENSHIFT_OPERATORS_NS_NAME) && checlusters.length === 0) {
```
Comment (AndrienkoAleksandr) [inline src/tasks/installers/olm.ts:336]: > kube.operatorSubscriptionExists(SUBSCRIPTION_NAME, DEFAULT_OPENSHIFT_OPERATORS_NS_NAME) && checlusters.length === 0

Maybe this check should be moved to the separated method, cause it was used not once?

File: src/tasks/installers/olm.ts [hunk 9]
```diff
@@ -313,43 +318,61 @@ export class OLMTasks {
...
+            await kube.deleteOperatorSubscription(SUBSCRIPTION_NAME, ctx.operatorNamespace)
```
Comment (AndrienkoAleksandr) [inline src/tasks/installers/olm.ts:337]: Sorry, but you completely removed case with delete operator subscription for one specific namespace. The same for another delete tasks....

File: src/commands/server/deploy.ts [hunk 10]
```diff
@@ -296,6 +296,10 @@ export default class Deploy extends Command {
         this.error(`🛑 The specified installer ${flags.installer} does not support Minishift`)
       }
 
+      if (flags['olm-channel'] === STABLE_ALL_NAMESPACES_CHANNEL_NAME && (flags.platform === 'minikube' || flags.platform === 'k8s')) {
```
Comment (tolusha) [inline src/commands/server/deploy.ts:299]:
```suggestion
      if (flags['olm-channel'] === STABLE_ALL_NAMESPACES_CHANNEL_NAME && isKubernetesPlatformFamily(flags.platform)) {
```

Comment (flacatus) [reply src/commands/server/deploy.ts:299]: @tolusha Should not be openshift?

File: src/tasks/installers/olm.ts [hunk 11]
```diff
@@ -68,6 +68,8 @@ export class OLMTasks {
       },
       {
         title: 'Create operator group',
+        // Not need it to create operator group in openshift-operators ns
+        enabled: () => flags['olm-channel'] !== STABLE_ALL_NAMESPACES_CHANNEL_NAME,
```
Comment (tolusha) [inline src/tasks/installers/olm.ts:72]: accordingly to the comment we have to compare operator namespace

File: src/tasks/installers/olm.ts [hunk 12]
```diff
@@ -174,22 +179,22 @@ export class OLMTasks {
...
+          await kube.waitUntilOperatorIsInstalled(ctx.installPlanName, ctx.operatorNamespace)
           task.title = `${task.title}...done.`
         },
       },
       {
         title: 'Set custom operator image',
         enabled: () => flags['che-operator-image'],
         task: async (_ctx: any, task: any) => {
-          const csvList = await kube.getClusterServiceVersions(flags.chenamespace)
+          const csvList = await kube.getClusterServiceVersions(flags.chenamespaces)
```
Comment (tolusha) [inline src/tasks/installers/olm.ts:197]: Typo ?

File: src/tasks/installers/olm.ts [hunk 13]
```diff
@@ -142,20 +144,23 @@ export class OLMTasks {
...
+              const catalogSourceNamespace = flags['catalog-source-namespace'] || ctx.operatorNamespace
+              subscription = this.constructSubscription(SUBSCRIPTION_NAME, flags['package-manifest-name'], ctx.operatorNamespace, catalogSourceNamespace, flags['olm-channel'], ctx.sourceName, ctx.approvalStarategy, ctx.startingCSV)
             } else if (VersionHelper.isDeployingStableVersion(flags) || flags['olm-channel'] === OLM_STABLE_CHANNEL_NAME) {
               // stable Che CatalogSource
-              subscription = this.constructSubscription(SUBSCRIPTION_NAME, DEFAULT_CHE_OLM_PACKAGE_NAME, flags.chenamespace, ctx.defaultCatalogSourceNamespace, OLM_STABLE_CHANNEL_NAME, ctx.catalogSourceNameStable, ctx.approvalStarategy, ctx.startingCSV)
+              subscription = this.constructSubscription(SUBSCRIPTION_NAME, DEFAULT_CHE_OLM_PACKAGE_NAME, ctx.operatorNamespace, ctx.defaultCatalogSourceNamespace, OLM_STABLE_CHANNEL_NAME, ctx.catalogSourceNameStable, ctx.approvalStarategy, ctx.startingCSV)
+            } else if (flags['olm-channel'] === STABLE_ALL_NAMESPACES_CHANNEL_NAME && isOpenshiftPlatformFamily(flags.platform)) {
```
Comment (tolusha) [inline src/tasks/installers/olm.ts:158]:
```suggestion
            } else if (flags['olm-channel'] === STABLE_ALL_NAMESPACES_CHANNEL_NAME) {
```
We already checked it in deploy command

File: src/tasks/installers/olm.ts [hunk 14]
```diff
@@ -313,11 +318,25 @@ export class OLMTasks {
...
         },
       },
+      {
+        title: `Check if operator is installed in ${DEFAULT_OPENSHIFT_OPERATORS_NS_NAME} namespace`,
+        task: async (ctx: any, task: any) => {
+          if (await kube.operatorSubscriptionExists(SUBSCRIPTION_NAME, DEFAULT_OPENSHIFT_OPERATORS_NS_NAME)) {
+            ctx.operatorNamespace = DEFAULT_OPENSHIFT_OPERATORS_NS_NAME
+            task.title = `${task.title}...Found`
+          }
+          ctx.operatorNamespace = flags.chenamespace
```
Comment (tolusha) [inline src/tasks/installers/olm.ts:328]:
```suggestion
          ctx.operatorNamespace = flags.chenamespace || DEFAULT_CHE_NAMESPACE 
```

File: src/tasks/installers/olm.ts [hunk 15]
```diff
@@ -313,11 +318,25 @@ export class OLMTasks {
           task.title = `${task.title}: ${ctx.isPreInstalledOLM}...OK`
         },
       },
+      {
+        title: `Check if operator is installed in ${DEFAULT_OPENSHIFT_OPERATORS_NS_NAME} namespace`,
+        task: async (ctx: any, task: any) => {
+          if (await kube.operatorSubscriptionExists(SUBSCRIPTION_NAME, DEFAULT_OPENSHIFT_OPERATORS_NS_NAME)) {
+            ctx.operatorNamespace = DEFAULT_OPENSHIFT_OPERATORS_NS_NAME
+            task.title = `${task.title}...Found`
```
Comment (tolusha) [inline src/tasks/installers/olm.ts:326]: `return` or `else`

File: src/tasks/installers/olm.ts (hunk 8 above)
Comment (tolusha) [reply src/tasks/installers/olm.ts:337]: pls revert

**Review by tolusha — APPROVED** [review]

**Pull request merged**